protocols/raft/src/main/java/io/atomix/protocols/raft/storage/snapshot/SnapshotStore.java

```diff
@@ -106,6 +109,22 @@ private void open() {
     }
   }
 
+  /**
+   * Returns the snapshot for the given primitive at the given index.
+   *
+   * @param serviceId the primitive for which to lookup the snapshot
+   * @param index     the index for which to lookup the snapshot
+   * @return the snapshot for the given service at the given index or {@code null} if the snapshot doesn't exist
+   */
+  public Snapshot getSnapshot(PrimitiveId serviceId, long index) {
+    Collection<Snapshot> snapshots = indexSnapshots.get(index);
+    return snapshots == null ? null :
+        snapshots.stream()
+            .filter(s -> s.serviceId().equals(serviceId))
```

Comment:
sorted?

Reply:
The set ensures there's only one snapshot per service so `findFirst` or `findAny` is fine here. Sorting would just be unnecessary overhead. The reason the collection is sorted elsewhere is so the leader can get a consistent list of snapshots to iterate through at a given index when replicating snapshots since multiple snapshots for different state machines can exist at a single index.